From 06e396564afd3488f44a17db96042121e6b8bb44 Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Wed, 28 Mar 2018 17:49:19 -0700
Subject: [PATCH 0337/1851] drm: xlnx: drv: Shutdown the drm device properly

commit fa529ebd64dc534dbc888586e473a15a6c6af3bb from
https://github.com/Xilinx/linux-xlnx.git

If a drm device is released immediately in shutdown callback by calling
drm_put_device(), it gives a warnning as below because there are some
resources, drm gem objects, that are still used by fbdev emulation and
attached to the drm device.

> reboot

or

> shutdown -h now

...
[   40.731016] [<ffffff800850adc0>] drm_gem_destroy+0x18/0x30
[   40.736485] [<ffffff800850d690>] drm_dev_fini+0x28/0x90
[   40.741693] [<ffffff800850d764>] drm_dev_unref+0x6c/0x78
[   40.746988] [<ffffff800850d7ac>] drm_put_dev+0x3c/0x70
[   40.752111] [<ffffff80085379b4>] xlnx_shutdown+0x14/0x20
[   40.757407] [<ffffff8008547b2c>] device_shutdown+0x114/0x1f0
[   40.763050] [<ffffff80080ba464>] kernel_restart_prepare+0x34/0x40
[   40.769125] [<ffffff80080ba544>] kernel_restart+0x14/0x78
[   40.774507] [<ffffff80080ba7e4>] SyS_reboot+0xbc/0x1e8
[   40.779627] Exception stack(0xffffff8013b13ec0 to 0xffffff8013b14000)
...

This fixes the warning by shutting down the drm device in complete
sequence.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_drv.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/xlnx/xlnx_drv.c b/drivers/gpu/drm/xlnx/xlnx_drv.c
index de9bfcaba6fd..e162bc85c34c 100644
--- a/drivers/gpu/drm/xlnx/xlnx_drv.c
+++ b/drivers/gpu/drm/xlnx/xlnx_drv.c
@@ -417,9 +417,7 @@ static int xlnx_platform_remove(struct platform_device *pdev)
 
 static void xlnx_platform_shutdown(struct platform_device *pdev)
 {
-	struct xlnx_drm *xlnx_drm = platform_get_drvdata(pdev);
-
-	drm_put_dev(xlnx_drm->drm);
+	component_master_del(&pdev->dev, &xlnx_master_ops);
 }
 
 static int __maybe_unused xlnx_pm_suspend(struct device *dev)
-- 
2.31.1

